### ISSUE TITLE
Discussion on Bell Polynomials

### ISSUE THREAD

**Issue by Nolord** [sagemath/sage#37727]
### Problem Description

The sage command for (partial) [Bell Polynomials](https://en.wikipedia.org/wiki/Bell_polynomials) `bell_polynomial(n,k)` outputs a polynomials in `n-k+1` variables. For that matter, the implementation begins with defining a ring [`R = PolynomialRing(ZZ, 'x', n - k + 1)`](https://github.com/sagemath/sage/blob/develop/src/sage/combinat/combinat.py#L2977) .

Here are some suggestions:
- If `n = k`, only one variable `x` is defined, which is inconsistent with the other cases where variables `x0` through `xn-k` would be defined. Which makes it impossible to manipulate different bell polynomials together (which happens in multiple scenarii) without substituting `x` by `x0` prior on `bell_polynomial(n,n)`.
- I'd like to [resuggest](https://github.com/sagemath/sage/issues/18338#issuecomment-1417683627) the idea of starting the indices of the variables at 1 instead of 0, just like on [wikipedia](https://en.wikipedia.org/wiki/Bell_polynomials). It is possible to add an additional optional variable to be able to choose the starting index to include [both conventions](https://github.com/sagemath/sage/issues/18338#issuecomment-1417683631).
- Functions for Complete Bell polynomials, (partial) _ordinary_ Bell polynomials and complete _ordinary_ Bell polynomials would be helpful.
- This might imply to also modify other multivariate polynomials for consistency.

### Proposed Solution

- [`R = PolynomialRing(ZZ, 'x', n - k + 1)`](https://github.com/sagemath/sage/blob/develop/src/sage/combinat/combinat.py#L2977) could be replaced by `R = PolynomialRing(ZZ, 'x0' if n == k else 'x', n - k + 1)` for the first suggestion.
- The second suggestion _would_ require an [ugly modification](https://github.com/sagemath/sage/issues/18338#issuecomment-1417683629).
```Python
def bell_polynomial(n: Integer, k: Integer, offset=False):
   R = PolynomialRing(ZZ, ['x%s'%(i+offset) for i in range(n-k+1)])
   ...
```
- The variant Bell polynomials have [simple formulas](https://en.wikipedia.org/wiki/Bell_polynomials) in terms of the partial Bell polynomials. The formula for the complete Bell polynomial is an example that illustrate the relevancy of the first suggestion :
```Python
def complete_bell_polynomial(n: Integer):
   return sum([bell_polynomial(n, k) for k in range(n+1)])
```
(though computationaly faster implementations might exist)
### Alternatives Considered

Modifying `PolynomialRing` itself?

### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

**Comment by Nolord** [comment]
> I am a strong -1 on changing the starting index. It's a variable name, the natural programming starting index (all the `R.gen(i)` would give xi+1, and not worth it.

I understand your point of view, but the motivation of starting at index one stems from the same root as your point. One way to define Bell polynomials is to first consider a sequence of coefficients $x = (x_n)_{n\in\mathbb N}$ for which we _impose_ $x_0 = 0$ and $x_1 \neq 0$ such that the power series $f(z) = \sum\_{n\geq0} x_n z^n/n!$ has compositional inverse. Then Bell polynomials can be defined by $f(z)^k/k! = \sum\_{n\geq k} B\_{n,k}(x) z^n/n!$ where $B\_{n,k}(x)$ depends on the sequence $x$ as a whole, but really depends only on a finite amount of them, that is $x\_1, \ldots, x\_{n-k+1}$. My point is, in the mathematical point of view, we do also index at $0$, but in that specific case, the dependance in $x_0$ is lost, that's why it seems that [Wikipedia](https://en.wikipedia.org/wiki/Bell_polynomials) chose the convention to index from $1$, but it really is from $0$.

> The first one for using `x0` is a reasonable special case. Adding the complete (exponential) Bell polynomials would be good as well. Perhaps with an optional argument to `bell_polynomial` (and change the default `k=None`). Please submit a PR for these.

I submitted the PR for my first suggestion as can be seen above.

Concerning your suggestion on implementing the complete Bell polynomial, is it really appropriate to use the same function `bell_polynomial` for it? As they remain separate objects, only linked by their name and by a formula. (as a reminder, complete Bell polynomials are defined by $\exp(f(z)) = \sum\_{n\geq0} B\_n(x_1, \ldots, x_n) z^n/n!$)

**Comment by tscrim** [comment]
> > I am a strong -1 on changing the starting index. It's a variable name, the natural programming starting index (all the `R.gen(i)` would give xi+1, and not worth it.
> 
> I understand your point of view, but the motivation of starting at index one stems from the same root as your point. One way to define Bell polynomials is to first consider a sequence of coefficients x=(xn)n∈N for which we _impose_ x0=0 and x1≠0 such that the power series f(z)=∑n≥0xnzn/n! has compositional inverse. Then Bell polynomials can be defined by f(z)k/k!=∑n≥kBn,k(x)zn/n! where Bn,k(x) depends on the sequence x as a whole, but really depends only on a finite amount of them, that is x1,…,xn−k+1. My point is, in the mathematical point of view, we do also index at 0, but in that specific case, the dependance in x0 is lost, that's why it seems that [Wikipedia](https://en.wikipedia.org/wiki/Bell_polynomials) chose the convention to index from 1, but it really is from 0.

We are within the context of programming here, and since Python (and subsequently Sage) made the choice to index everything from `0`, we should respect that as much as possible. it is much easier to make an indexing shift in the mathematical formulation. To this point, if I was looking at the variable set on its own, I would want to run the variables $x_0, \ldots, x_{n-k}$, but that might be because of my biases. (I don't see what the compositional inverse has to do with anything.)

At the end of the day, we have to weight the costs of the convention choices. Yes, there is a slight danger of confusion and the shifts might be more annoying, but the internal code would become much more complicated, we would break backwards compatibility (without sufficient justification IMO), and users would have a harder time working with the result (i.e., it would diverge from the natural construction `PolynomialRing(QQ, 'x', n)`).

> > The first one for using `x0` is a reasonable special case. Adding the complete (exponential) Bell polynomials would be good as well. Perhaps with an optional argument to `bell_polynomial` (and change the default `k=None`). Please submit a PR for these.

> Concerning your suggestion on implementing the complete Bell polynomial, is it really appropriate to use the same function `bell_polynomial` for it? As they remain separate objects, only linked by their name and by a formula. (as a reminder, complete Bell polynomials are defined by $\exp⁡(f(z)) = \sum_{n\geq0} B_n(x_1,\ldots,x_n) z_n / n!$)

Yes; the names coincide, they are clearly linked, and it makes it easily discoverable. I would just do
```python
def bell_polynomial(n, k=None):
    if k is None:
        return sum(bell_polynomial(n, k) for k in range(n+1))
    [rest of the Bell polynomial code
```
and just indicate in the docstring what not passing `k` means.

----

Also, permit me to do a shameless plug of working with power series constructions in Sage:
```
sage: R.<x> = InfinitePolynomialRing(QQ)
sage: L.<z> = LazyPowerSeriesRing(R)
sage: f = L(lambda n: x[n-1], valuation=1)
sage: ef = exp(f)
sage: [ef[n] * factorial(n) for n in range(7)]
[1,
 x_0,
 2*x_1 + x_0^2,
 6*x_2 + 6*x_1*x_0 + x_0^3,
 24*x_3 + 24*x_2*x_0 + 12*x_1^2 + 12*x_1*x_0^2 + x_0^4,
 120*x_4 + 120*x_3*x_0 + 120*x_2*x_1 + 60*x_2*x_0^2 + 60*x_1^2*x_0 + 20*x_1*x_0^3 + x_0^5,
 720*x_5 + 720*x_4*x_0 + 720*x_3*x_1 + 360*x_3*x_0^2 + 360*x_2^2 + 720*x_2*x_1*x_0 + 120*x_2*x_0^3 + 120*x_1^3 + 180*x_1^2*x_0^2 + 30*x_1*x_0^4 + x_0^6]
sage: f4 = f^4 / factorial(4)
sage: [f4[n] * factorial(n) for n in range(4, 9)]
[x_0^4,
 20*x_1*x_0^3,
 120*x_2*x_0^3 + 180*x_1^2*x_0^2,
 840*x_3*x_0^3 + 2520*x_2*x_1*x_0^2 + 840*x_1^3*x_0,
 6720*x_4*x_0^3 + 20160*x_3*x_1*x_0^2 + 10080*x_2^2*x_0^2 + 20160*x_2*x_1^2*x_0 + 1680*x_1^4]
```

**Comment by Nolord** [comment]
> If we wanted to be really good, they should be separate PRs with one a dependency on the other, but I am okay doing it all on #37733 (i.e., on that branch; of course updating the PR description too).
Will do so.
> It might also be good to add an additional optional argument for the exponential Bell polynomials.
An additional optional argument to do what exactly?

**Comment by tscrim** [comment]
> > It might also be good to add an additional optional argument for the exponential Bell polynomials.
> 
> An additional optional argument to do what exactly?

To rescale the variables $x_i \mapsto (i+1) x_i$ (with the $x_0, \ldots, x_{n-k}$ convention) with an extra overall $k!/n!$ factor. That is, to match the definitions you gave (compare the output of the power series code with `bell_polynomial()`).

**Comment by tscrim** [comment]
Yes and yes. Although the name is confusing to me (as a non-expert on these polynomials) compared with the complete Bell polynomials, which aren't coming from the exponentiation you mentioned.
